### ISSUE TITLE
[Core/Cpp]Empty Ray cpp_worker_command

### ISSUE THREAD

**Issue by wilsonwang371** [ray-project/ray#31477]
### What happened + What you expected to happen

I am working on Ray Cpp worker and I have problem running the example in cpp/example/run.sh

Initially, the error looks like that the cpp worker registration failure.
Note: I modified the Ray code to print out the exact command when starting a new process.
```
$ ./run.sh 
INFO: Analyzed target //:example (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:example up-to-date:
  bazel-bin/example
INFO: Elapsed time: 0.395s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
[2023-01-05 10:43:32,037 I 54352 9035610] config_internal.cc:192: No code search path found yet. The program location path "/Users/xiaoshuang/Documents/GitHub/ray-inf/cpp/example/bazel-bin" will be added for searching dynamic libraries by default. And you can add some search paths by '--ray_code_search_path'
[2023-01-05 10:43:32,038 I 54352 9035610] process_helper.cc:45: ray start --head --port 6379 --redis-password 5241590000000000 --node-ip-address '192.168.1.151'
Usage stats collection is enabled. To disable this, add `--disable-usage-stats` to the command that starts the cluster, or run the following command: `ray disable-usage-stats` before starting the cluster. See https://docs.ray.io/en/master/cluster/usage-stats.html for more details.

Local node IP: 192.168.1.151
Launch Process: '/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/core/src/ray/gcs/gcs_server' '--log_dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/logs' '--config_list=eyJvYmplY3Rfc3BpbGxpbmdfY29uZmlnIjogIntcInR5cGVcIjogXCJmaWxlc3lzdGVtXCIsIFwicGFyYW1zXCI6IHtcImRpcmVjdG9yeV9wYXRoXCI6IFwiL3RtcC9yYXkvc2Vzc2lvbl8yMDIzLTAxLTA1XzEwLTQzLTMzXzcwMjcwOF81NDM1NFwifX0iLCAiaXNfZXh0ZXJuYWxfc3RvcmFnZV90eXBlX2ZzIjogdHJ1ZX0=' '--gcs_server_port=6379' '--metrics-agent-port=59585' '--node-ip-address=192.168.1.151' '--session-name=session_2023-01-05_10-43-33_702708_54354' '--redis_password=5241590000000000'
Launch Process: '/Users/xiaoshuang/opt/miniconda3/bin/python' '-u' '/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/autoscaler/_private/monitor.py' '--logs-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/logs' '--logging-rotate-bytes=536870912' '--logging-rotate-backup-count=5' '--gcs-address=192.168.1.151:6379' '--monitor-ip=192.168.1.151'
Launch Process: '/Users/xiaoshuang/opt/miniconda3/bin/python' '/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/_private/workers/setup_worker.py' '-m' 'ray.util.client.server' '--address=192.168.1.151:6379' '--host=0.0.0.0' '--port=10001' '--mode=proxy' '--language=PYTHON' '--redis-password=5241590000000000' '--metrics-agent-port=59585'
Launch Process: '/Users/xiaoshuang/opt/miniconda3/bin/python' '/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/dashboard/dashboard.py' '--host=localhost' '--port=8265' '--port-retries=0' '--temp-dir=/tmp/ray' '--log-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/logs' '--session-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354' '--logging-rotate-bytes=536870912' '--logging-rotate-backup-count=5' '--gcs-address=192.168.1.151:6379' '--minimal'
Launch Process: '/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/core/src/ray/raylet/raylet' '--raylet_socket_name=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/sockets/raylet' '--store_socket_name=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/sockets/plasma_store' '--object_manager_port=0' '--min_worker_port=10002' '--max_worker_port=19999' '--node_manager_port=0' '--node_ip_address=192.168.1.151' '--maximum_startup_concurrency=8' '--static_resource_list=node:192.168.1.151,1.0,CPU,8,memory,4113336731,object_store_memory,2056668364' '--python_worker_command=/Users/xiaoshuang/opt/miniconda3/bin/python /Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/_private/workers/setup_worker.py /Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/_private/workers/default_worker.py --node-ip-address=192.168.1.151 --node-manager-port=RAY_NODE_MANAGER_PORT_PLACEHOLDER --object-store-name=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/sockets/plasma_store --raylet-name=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/sockets/raylet --redis-address=None --temp-dir=/tmp/ray --metrics-agent-port=59585 --logging-rotate-bytes=536870912 --logging-rotate-backup-count=5 --gcs-address=192.168.1.151:6379 --session-name=session_2023-01-05_10-43-33_702708_54354 --temp-dir=/tmp/ray --webui= --storage=None RAY_WORKER_DYNAMIC_OPTION_PLACEHOLDER --redis-password=5241590000000000' '--java_worker_command=' '--cpp_worker_command=' '--native_library_path=/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/cpp/lib' '--temp_dir=/tmp/ray' '--session_dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354' '--log_dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/logs' '--resource_dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/runtime_resources' '--metrics-agent-port=59585' '--metrics_export_port=62137' '--object_store_memory=2056668364' '--plasma_directory=/tmp' '--ray-debugger-external=0' '--gcs-address=192.168.1.151:6379' '--session-name=session_2023-01-05_10-43-33_702708_54354' '--agent_command=/Users/xiaoshuang/opt/miniconda3/bin/python -u /Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/dashboard/agent.py --node-ip-address=192.168.1.151 --metrics-export-port=62137 --dashboard-agent-port=59585 --listen-port=52365 --node-manager-port=RAY_NODE_MANAGER_PORT_PLACEHOLDER --object-store-name=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/sockets/plasma_store --raylet-name=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/sockets/raylet --temp-dir=/tmp/ray --session-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354 --runtime-env-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/runtime_resources --log-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/logs --logging-rotate-bytes=536870912 --logging-rotate-backup-count=5 --session-name=session_2023-01-05_10-43-33_702708_54354 --gcs-address=192.168.1.151:6379 --minimal' '--node-name=192.168.1.151'
Launch Process: '/Users/xiaoshuang/opt/miniconda3/bin/python' '-u' '/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/_private/log_monitor.py' '--logs-dir=/tmp/ray/session_2023-01-05_10-43-33_702708_54354/logs' '--gcs-address=192.168.1.151:6379' '--logging-rotate-bytes=536870912' '--logging-rotate-backup-count=5'

--------------------
Ray runtime started.
--------------------

Next steps
  To connect to this Ray runtime from another node, run
    ray start --address='192.168.1.151:6379'
  
  Alternatively, use the following Python code:
    import ray
    ray.init(address='auto', _node_ip_address='192.168.1.151')
  
  To connect to this Ray runtime from outside of the cluster, for example to
  connect to a remote cluster from your laptop directly, use the following
  Python code:
    import ray
    ray.init(address='ray://<head_node_ip_address>:10001')
  
  To see the status of the cluster, use
    ray status
  
  If connection fails, check your firewall settings and network configuration.
  
  To terminate the Ray runtime, run
    ray stop
[2023-01-05 10:43:37,753 E 54352 9035610] core_worker.cc:182: Failed to register worker 01000000ffffffffffffffffffffffffffffffffffffffffffffffff to Raylet. IOError: [RayletClient] Unable to register worker with raylet. No such file or directory
```

If you search `--cpp_worker_command` you will see empty argument in previous command line log.

When I looked at the driver log, I can see this:

```
$ cat cpp-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_54352.log 
[2023-01-05 10:43:37,718 I 54352 9035610] core_worker_process.cc:107: Constructing CoreWorkerProcess. pid: 54352
[2023-01-05 10:43:37,720 I 54352 9035610] io_service_pool.cc:35: IOServicePool is running with 1 io_service.
[2023-01-05 10:43:37,753 E 54352 9035610] core_worker.cc:182: Failed to register worker 01000000ffffffffffffffffffffffffffffffffffffffffffffffff to Raylet. IOError: [RayletClient] Unable to register worker with raylet. No such file or directory
```

```
$ cat raylet.err 
[2023-01-05 10:43:37,749 C 54365 9035815] (raylet) worker_pool.cc:1389:  Check failed: state != states_by_lang_.end() Required Language isn't supported: CPP
*** StackTrace Information ***
0   raylet                              0x000000010ee88e61 _ZN3raylsERNSt3__113basic_ostreamIcNS0_11char_traitsIcEEEERKNS_10StackTraceE + 65 ray::operator<<()
1   raylet                              0x000000010ee8d012 _ZN3ray13SpdLogMessage5FlushEv + 114 ray::SpdLogMessage::Flush()
2   raylet                              0x000000010ee8cee2 _ZN3ray13SpdLogMessageD2Ev + 18 ray::SpdLogMessage::~SpdLogMessage()
3   raylet                              0x000000010ee8c2cf _ZN3ray6RayLogD2Ev + 47 ray::RayLog::~RayLog()
4   raylet                              0x000000010eb582e1 _ZN3ray6raylet10WorkerPool19GetStateForLanguageERKNS_3rpc8LanguageE + 433 ray::raylet::WorkerPool::GetStateForLanguage()
5   raylet                              0x000000010eb5b4d2 _ZN3ray6raylet10WorkerPool14RegisterDriverERKNSt3__110shared_ptrINS0_15WorkerInterfaceEEERKNS_3rpc9JobConfigENS2_8functionIFvNS_6StatusEiEEE + 578 ray::raylet::WorkerPool::RegisterDriver()
6   raylet                              0x000000010eaa4145 _ZN3ray6raylet11NodeManager35ProcessRegisterClientRequestMessageERKNSt3__110shared_ptrINS_16ClientConnectionEEEPKh + 2213 ray::raylet::NodeManager::ProcessRegisterClientRequestMessage()
7   raylet                              0x000000010eaa358a _ZN3ray6raylet11NodeManager20ProcessClientMessageERKNSt3__110shared_ptrINS_16ClientConnectionEEExPKh + 1066 ray::raylet::NodeManager::ProcessClientMessage()
8   raylet                              0x000000010eb42ed9 _ZNSt3__110__function6__funcIZN3ray6raylet6Raylet12HandleAcceptERKN5boost6system10error_codeEE3$_2NS_9allocatorISA_EEFvNS_10shared_ptrINS2_16ClientConnectionEEExRKNS_6vectorIhNSB_IhEEEEEEclEOSF_OxSK_ + 57 std::__1::__function::__func<>::operator()()
9   raylet                              0x000000010edb5070 _ZN3ray16ClientConnection14ProcessMessageERKN5boost6system10error_codeE + 1024 ray::ClientConnection::ProcessMessage()
10  raylet                              0x000000010edc3f07 _ZN12EventTracker15RecordExecutionERKNSt3__18functionIFvvEEENS0_10shared_ptrI11StatsHandleEE + 87 EventTracker::RecordExecution()
11  raylet                              0x000000010edbfd6d _ZN5boost4asio6detail7read_opINS0_19basic_stream_socketINS0_7generic15stream_protocolENS0_9execution12any_executorIJNS6_12context_as_tIRNS0_17execution_contextEEENS6_6detail8blocking7never_tILi0EEENS6_11prefer_onlyINSD_10possibly_tILi0EEEEENSG_INSC_16outstanding_work9tracked_tILi0EEEEENSG_INSK_11untracked_tILi0EEEEENSG_INSC_12relationship6fork_tILi0EEEEENSG_INSR_14continuation_tILi0EEEEEEEEEENS0_17mutable_buffers_1EPKNS0_14mutable_bufferENS1_14transfer_all_tEZN3ray16ClientConnection20ProcessMessageHeaderERKNS_6system10error_codeEE3$_7EclES1A_mi + 573 boost::asio::detail::read_op<>::operator()()
12  raylet                              0x000000010edc0146 _ZN5boost4asio6detail23reactive_socket_recv_opINS0_17mutable_buffers_1ENS1_7read_opINS0_19basic_stream_socketINS0_7generic15stream_protocolENS0_9execution12any_executorIJNS8_12context_as_tIRNS0_17execution_contextEEENS8_6detail8blocking7never_tILi0EEENS8_11prefer_onlyINSF_10possibly_tILi0EEEEENSI_INSE_16outstanding_work9tracked_tILi0EEEEENSI_INSM_11untracked_tILi0EEEEENSI_INSE_12relationship6fork_tILi0EEEEENSI_INST_14continuation_tILi0EEEEEEEEEES3_PKNS0_14mutable_bufferENS1_14transfer_all_tEZN3ray16ClientConnection20ProcessMessageHeaderERKNS_6system10error_codeEE3$_7EES10_E11do_completeEPvPNS1_19scheduler_operationES1B_m + 534 boost::asio::detail::reactive_socket_recv_op<>::do_complete()
13  raylet                              0x000000010f4957d2 _ZN5boost4asio6detail9scheduler10do_run_oneERNS1_27conditionally_enabled_mutex11scoped_lockERNS1_21scheduler_thread_infoERKNS_6system10error_codeE + 706 boost::asio::detail::scheduler::do_run_one()
14  raylet                              0x000000010f48be41 _ZN5boost4asio6detail9scheduler3runERNS_6system10error_codeE + 225 boost::asio::detail::scheduler::run()
15  raylet                              0x000000010f48bd4b _ZN5boost4asio10io_context3runEv + 43 boost::asio::io_context::run()
16  raylet                              0x000000010ea4a49f main + 3263 main
17  dyld                                0x000000011ff0552e start + 462 start

/Users/xiaoshuang/opt/miniconda3/lib/python3.10/site-packages/ray-3.0.0.dev0-py3.10-macosx-10.9-x86_64.egg/ray/dashboard/agent.py:51: DeprecationWarning: There is no current event loop
  aiogrpc.init_grpc_aio()
```

I did some investigation and found that the possible reason for this failure is because of this:

In `python/ray/_private/services.py` this line is actually
```
DEFAULT_WORKER_EXECUTABLE = os.path.join(RAY_PATH, "cpp", "default_worker" + EXE_SUFFIX)
```

But the actual `DEFAULT_WORKER_EXECUTABLE` do not exists on my mac.

The `default_worker` is not in `ray` python package directory. It is actually inside `ray-cpp` python package directory.

### Versions / Dependencies

MacOS Monterey 12.6 
Python 3.10.8
Ray branch: master, commit fba15f6a85e4809e9ef64970491e6039bdb2b0fa

### Reproduction script

n/a

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by wilsonwang371** [comment]
This might be related to https://github.com/ray-project/ray/pull/31312
FYI @iycheng @simon-mo @rkooo567

**Comment by iycheng** [comment]
@wilsonwang371 I merged https://github.com/ray-project/ray/pull/31312 could you give it another try?

**Comment by glennmoy** [comment]
Just to note that I'm having this exact same issue while just trying to `init` a ray session using the CPP API on linux x86_64.
I working on a branch forked from 27bfa89dc0f656208280e7d3fef37c288cdae232.

For reference, I also reran the OP's MWE and see the exact same errors in the exact same files and the fix in https://github.com/ray-project/ray/pull/31312 doesn't seem to affect things whatsoever (I also ran against 27bfa89dc0f656208280e7d3fef37c288cdae232 to check).

@wilsonwang371 did the fix above work for your or did you have another workaround? This is the only hit on google for this error so I'm hoping there's a solution! thanks

**Comment by wilsonwang371** [comment]
> Just to note that I'm having this exact same issue while just trying to `init` a ray session using the CPP API on linux x86_64. I working on a branch forked from [3e4450a](https://github.com/ray-project/ray/commit/3e4450a640012e927644ad353f204543edb941ff).
> 
> For reference, I also reran the OP's MWE and see the exact same errors in the exact same files and the fix in #31312 doesn't seem to fix the problem on my machine (I also ran against [27bfa89](https://github.com/ray-project/ray/commit/27bfa89dc0f656208280e7d3fef37c288cdae232) to check).
> 
> @wilsonwang371 did the fix above work for your or did you have another workaround? This is the only issue I found for this error (besides #28200) so I'm hoping there's a solution! thanks

Did you build the binaries all by yourself. I think my final solution was to build the wheel files instead of building egg files.

**Comment by glennmoy** [comment]
> Contributor

yeah I built the binaries (`core_worker_lib`, etc.) myself (as I'm linking them to another C++ library)  using `bazel-5.4.1`.

>Basically, I build the binaries using this command instead of using python setup.py.

Thanks - the build worked (after installing `python3-dev`) but the `run.sh` script is now segfaulting. I think this is a separate issue so I won't go into details here.

```
./run.sh: line 13: 1014416 Segmentation fault      LD_LIBRARY_PATH="$ROOT_DIR/thirdparty/lib" "${ROOT_DIR}"/bazel-bin/example
```

My custom library has also progressed so thanks for your help! Do you have an understanding of why that worked? I'm still getting used to Ray/Bazel so would be keen to know.

**Comment by wilsonwang371** [comment]
> > Did you build the binaries all by yourself. I think my final solution was to build the wheel files instead of building egg files.
> 
> yeah I built the binaries (`core_worker_lib`, etc.) myself (as I'm linking them to another C++ library) using `bazel-5.4.1`.
> 
> > Basically, I build the binaries using this command instead of using python setup.py.
> 
> Thanks - the build worked (after installing `python3-dev`) but the `run.sh` script is now segfaulting. I think this is a separate issue so I won't go into details here.
> 
> ```
> ./run.sh: line 13: 1014416 Segmentation fault      LD_LIBRARY_PATH="$ROOT_DIR/thirdparty/lib" "${ROOT_DIR}"/bazel-bin/example
> ```
> 
> My custom library has also progressed so thanks for your help! Do you have an understanding of why that worked? I'm still getting used to Ray/Bazel so would be keen to know.

Long story short, eggs and wheels install your binary to different directories.
